# EDA


- The os module has a perfect method to list files in a directory.
- Pandas json normalize could work here but is not necessary to convert the JSON data to a dataframe.
- You may need a nested for-loop to access each sale!
- We've put a lot of time into creating the structure of this repository, and it's a good example for future projects.  In the file functions_variables.py, there is an example function that you can import and use.  If you have any variables, functions or classes that you want to make, they can be put in the functions_variables.py file and imported into a notebook.  Note that only .py files can be imported into a notebook. If you want to import everything from a .py file, you can use the following:
```python
from functions_variables import *
```
If you just import functions_variables, then each object from the file will need to be prepended with "functions_variables"\
Using this .py file will keep your notebooks very organized and make it easier to reuse code between notebooks.

In [28]:
# (this is not an exhaustive list of libraries)
import pandas as pd
import numpy as np
import os
import json
from pprint import pprint
from functions_variables import encode_tags
from os import listdir
from os.path import join, isfile


## Data Importing

In [29]:
# load one file first to see what type of data you're dealing with and what attributes it has

In [30]:
# loop over all files and put them into a dataframe
import os
import pandas as pd

# Directory containing your JSON files
json_directory = 'C:/Users/spencer.fargey/LighthouseLabs/tdsf-midterm/data'  # Replace with the actual path to your directory

# Initialize an empty list to store DataFrames from each JSON file
dataframes = []

# Function to load and process each JSON file
def load_json_file(fname):
    """
    This function loads a JSON file (assumed to be ASCII-readable)
    and extracts the 'results' from the 'data' field.
    """
    with open(fname, 'r') as f:
        try:
            data_json = json.load(f)
            # Check if 'data' and 'results' exist in the loaded JSON
            if 'data' in data_json and 'results' in data_json['data']:
                df = pd.json_normalize(data_json['data']['results'])
            else:
                raise KeyError("'data' or 'results' field not found in the JSON file")
        except json.JSONDecodeError as e:
            print(f"Error decoding JSON in file {fname}: {e}")
            return None
        except KeyError as e:
            print(f"Key error in file {fname}: {e}")
            return None
    
    return df

# Loop through all files in the directory
for filename in os.listdir(json_directory):
    if filename.endswith(".json"):  # Check if the file is a JSON file
        file_path = os.path.join(json_directory, filename)  # Full file path
        df = load_json_file(file_path)  # Call load_json_file with fname
        if df is not None:  # If the DataFrame is valid, append to list
            dataframes.append(df)

# Combine all DataFrames into one
if dataframes:
    final_df = pd.concat(dataframes, ignore_index=True)
    print(final_df)
else:
    print("No valid JSON files were processed.")


FileNotFoundError: [Errno 2] No such file or directory: 'C:/Users/spencer.fargey/LighthouseLabs/tdsf-midterm/data'

In [ ]:
# Save the first 50 rows of the dataframe to a CSV file
output_csv_path = 'C:/Users/spencer.fargey/LighthouseLabs/tdsf-midterm/Housing_Data.csv'
final_df.to_csv(output_csv_path, index=False)

print(f"CSV saved to: {output_csv_path}")


CSV saved to: C:/Users/spencer.fargey/LighthouseLabs/tdsf-midterm/Housing_Data.csv


## Data Cleaning and Wrangling

At this point, ensure that you have all sales in a dataframe.
- Take a quick look at your data (i.e. `.info()`, `.describe()`) - what do you see?
- Is each cell one value, or do some cells have lists?
- What are the data types of each column?
- Some sales may not actually include the sale price (target).  These rows should be dropped.
- There are a lot of NA/None values.  Should these be dropped or replaced with something?
    - You can drop rows or use various methods to fills NA's - use your best judgement for each column 
    - i.e. for some columns (like Garage), NA probably just means no Garage, so 0
- Drop columns that aren't needed
    - Don't keep the list price because it will be too close to the sale price. Assume we want to predict the price of houses not yet listed

In [61]:
# load data here, data already concatenated into the csv in previous step

df = pd.read_csv("/Users/thomasdoherty/Desktop/tdsf-midterm/Housing_Data.csv") # loading the data from the csv file compiled from the json files above

df.head(5) # make sure it's loaded correctly

last_update_date                                               tags  \
0  2023-09-19T20:52:50Z  ['carport', 'community_outdoor_space', 'cul_de...   
1                   NaN                                                NaN   
2                   NaN                                                NaN   
3                   NaN                                                NaN   
4                   NaN                                                NaN   

                                           permalink status  \
0       9453-Herbert-Pl_Juneau_AK_99801_M90744-30767   sold   
1  8477-Thunder-Mountain-Rd_Juneau_AK_99801_M9424...   sold   
2      4515-Glacier-Hwy_Juneau_AK_99801_M94790-68516   sold   
3  17850-Point-Stephens-Rd_Juneau_AK_99801_M98793...   sold   
4  9951-Stephen-Richards-Memorial-Dr_Juneau_AK_99...   sold   

                     list_date  open_houses  \
0  2023-06-29T21:16:25.000000Z          NaN   
1                          NaN          NaN   
2                          NaN          NaN   
3                          NaN          NaN   
4                          NaN          NaN   

                                            branding  list_price  \
0  [{'name': 'EXP Realty LLC - Southeast Alaska',...    554950.0   
1  [{'name': None, 'photo': None, 'type': 'Office'}]         NaN   
2  [{'name': None, 'photo': None, 'type': 'Office'}]         NaN   
3  [{'name': None, 'photo': None, 'type': 'Office'}]         NaN   
4  [{'name': None, 'photo': None, 'type': 'Office'}]         NaN   

    property_id                                             photos  ...  \
0  9.074431e+09  [{'tags': [{'label': 'house_view', 'probabilit...  ...   
1  9.424984e+09                                                NaN  ...   
2  9.479069e+09                                                NaN  ...   
3  9.879332e+09                                                NaN  ...   
4  9.521640e+09                                                NaN  ...   

   location.county.fips_code location.county.name  primary_photo  source  \
0                        NaN               Juneau            NaN     NaN   
1                        NaN               Juneau            NaN     NaN   
2                        NaN               Juneau            NaN     NaN   
3                        NaN               Juneau            NaN     NaN   
4                        NaN               Juneau            NaN     NaN   

  products location.address.coordinate  other_listings community.advertisers  \
0      NaN                         NaN             NaN                   NaN   
1      NaN                         NaN             NaN                   NaN   
2      NaN                         NaN             NaN                   NaN   
3      NaN                         NaN             NaN                   NaN   
4      NaN                         NaN             NaN                   NaN   

  community.description.name location.county  
0                        NaN             NaN  
1                        NaN             NaN  
2                        NaN             NaN  
3                        NaN             NaN  
4                        NaN             NaN  

[5 rows x 67 columns]

In [62]:
# basic exploration of the df

df.shape # 67 possible variables and 8.2k records

(8191, 67)

In [63]:
df.info() # lots of missing values, some completely full columns, some completely empty columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8191 entries, 0 to 8190
Data columns (total 67 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   last_update_date                       8125 non-null   object 
 1   tags                                   7638 non-null   object 
 2   permalink                              8159 non-null   object 
 3   status                                 8159 non-null   object 
 4   list_date                              7752 non-null   object 
 5   open_houses                            0 non-null      float64
 6   branding                               8159 non-null   object 
 7   list_price                             7721 non-null   float64
 8   property_id                            8159 non-null   float64
 9   photos                                 7403 non-null   object 
 10  community                              0 non-null      float64
 11  virt

From just looking at the info above there are **lots of null values**. There doesn't seem any point in getting descriptive statistics until we remove all the completely empty numerical columns. 

Check all columns for completely null

In [65]:
total = df.shape[0] # total number of rows - 8191

# how many missing values are there in each column?
null_counts = df.isnull().sum() # there are many columns with 8191 nulls - the entire column is null

completely_null_cols = null_counts[null_counts == total].index.tolist()

df[completely_null_cols].head(10) # these columns are completely null

open_houses  community  description.name  description.baths_1qtr  \
0          NaN        NaN               NaN                     NaN   
1          NaN        NaN               NaN                     NaN   
2          NaN        NaN               NaN                     NaN   
3          NaN        NaN               NaN                     NaN   
4          NaN        NaN               NaN                     NaN   
5          NaN        NaN               NaN                     NaN   
6          NaN        NaN               NaN                     NaN   
7          NaN        NaN               NaN                     NaN   
8          NaN        NaN               NaN                     NaN   
9          NaN        NaN               NaN                     NaN   

   flags.is_new_construction  flags.is_for_rent  flags.is_subdivision  \
0                        NaN                NaN                   NaN   
1                        NaN                NaN                   NaN   
2                        NaN                NaN                   NaN   
3                        NaN                NaN                   NaN   
4                        NaN                NaN                   NaN   
5                        NaN                NaN                   NaN   
6                        NaN                NaN                   NaN   
7                        NaN                NaN                   NaN   
8                        NaN                NaN                   NaN   
9                        NaN                NaN                   NaN   

   flags.is_contingent  flags.is_pending  flags.is_plan  flags.is_coming_soon  \
0                  NaN               NaN            NaN                   NaN   
1                  NaN               NaN            NaN                   NaN   
2                  NaN               NaN            NaN                   NaN   
3                  NaN               NaN            NaN                   NaN   
4                  NaN               NaN            NaN                   NaN   
5                  NaN               NaN            NaN                   NaN   
6                  NaN               NaN            NaN                   NaN   
7                  NaN               NaN            NaN                   NaN   
8                  NaN               NaN            NaN                   NaN   
9                  NaN               NaN            NaN                   NaN   

   primary_photo  source  products  location.address.coordinate  \
0            NaN     NaN       NaN                          NaN   
1            NaN     NaN       NaN                          NaN   
2            NaN     NaN       NaN                          NaN   
3            NaN     NaN       NaN                          NaN   
4            NaN     NaN       NaN                          NaN   
5            NaN     NaN       NaN                          NaN   
6            NaN     NaN       NaN                          NaN   
7            NaN     NaN       NaN                          NaN   
8            NaN     NaN       NaN                          NaN   
9            NaN     NaN       NaN                          NaN   

   other_listings  location.county  
0             NaN              NaN  
1             NaN              NaN  
2             NaN              NaN  
3             NaN              NaN  
4             NaN              NaN  
5             NaN              NaN  
6             NaN              NaN  
7             NaN              NaN  
8             NaN              NaN  
9             NaN              NaN

In [66]:
df[completely_null_cols].shape # there are 17 columns that are completely null

(8191, 17)

In [67]:
# drop the completely null columns
df = df.drop(columns=completely_null_cols)

In [68]:
df.shape # now is it 50 columns as the 17 completely null are gone

(8191, 50)

We have removed the 17 completely null columns (such as flag.is_subdivision, flag.is_for_rent, location.county...)
Some thoughts:
- The is_subdivision and is_for_rent might have been useful in some predictive way of sale price of a house and they would appear to be yes/no answers (e.g. it is a subdivision / for rent or it is not) but the fact there are no records with a yes (if we make the assumption that NaN means 0, or 'No') means the model wouldn't be able to get any predictive power from that column.

I will look at partially null numerical columns from here and then the categorical columns.

In [69]:
# how many features are numerical and how many are categorical?
numerical = df.select_dtypes(include=[np.number]).columns
categorical = df.select_dtypes(include=['object']).columns

df[numerical].head(10) # there are 37 numerical features

list_price   property_id    listing_id  price_reduced_amount  \
0    554950.0  9.074431e+09  2.957242e+09               45000.0   
1         NaN  9.424984e+09           NaN                   NaN   
2         NaN  9.479069e+09           NaN                   NaN   
3         NaN  9.879332e+09           NaN                   NaN   
4         NaN  9.521640e+09           NaN                   NaN   
5    415000.0  7.412310e+09  2.958925e+09                   NaN   
6    835000.0  7.660491e+09  2.958924e+09                   NaN   
7         NaN  8.274121e+09           NaN                   NaN   
8         NaN  9.424984e+09           NaN                   NaN   
9         NaN  9.479069e+09           NaN                   NaN   

   source.plan_id  description.year_built  description.baths_3qtr  \
0             NaN                  1963.0                     NaN   
1             NaN                     NaN                     NaN   
2             NaN                     NaN                     NaN   
3             NaN                     NaN                     NaN   
4             NaN                     NaN                     NaN   
5             NaN                  1969.0                     NaN   
6             NaN                  1920.0                     NaN   
7             NaN                  2002.0                     NaN   
8             NaN                     NaN                     NaN   
9             NaN                     NaN                     NaN   

   description.sold_price  description.baths_full  description.baths_half  \
0                     NaN                     2.0                     NaN   
1                     NaN                     NaN                     NaN   
2                     NaN                     NaN                     NaN   
3                     NaN                     NaN                     NaN   
4                     NaN                     NaN                     NaN   
5                     NaN                     1.0                     NaN   
6                     NaN                     3.0                     NaN   
7                     NaN                     2.0                     NaN   
8                     NaN                     NaN                     NaN   
9                     NaN                     NaN                     NaN   

   description.lot_sqft  description.sqft  description.baths  \
0               10454.0            1821.0                2.0   
1                   NaN               NaN                NaN   
2                   NaN               NaN                NaN   
3                   NaN               NaN                NaN   
4                   NaN               NaN                NaN   
5                   NaN             950.0                1.0   
6                5875.0            3860.0                3.0   
7                7476.0            1375.0                2.0   
8                   NaN               NaN                NaN   
9                   NaN               NaN                NaN   

   description.garage  description.stories  description.beds  \
0                 1.0                  NaN               3.0   
1                 NaN                  NaN               NaN   
2                 NaN                  NaN               NaN   
3                 NaN                  NaN               NaN   
4                 NaN                  NaN               NaN   
5                 NaN                  NaN               2.0   
6                 2.0                  NaN               5.0   
7                 NaN                  1.0               3.0   
8                 NaN                  NaN               NaN   
9                 NaN                  NaN               NaN   

   location.address.postal_code  location.address.coordinate.lon  \
0                       99801.0                      -134.593720   
1                       99801.0                              NaN   
2                       99801.0            

In [70]:
df[numerical].shape # there are 20 numerical features

(8191, 20)

In [75]:
# calculate the proportion of null entries in each column
null_proportion = df.isnull().sum() / total * 100

round(null_proportion.sort_values(ascending=False), 2)

community.description.name               99.94
source.plan_id                           99.94
source.spec_id                           99.94
community.advertisers                    99.94
flags.is_foreclosure                     99.49
description.baths_3qtr                   93.09
virtual_tours                            83.51
description.sub_type                     82.58
description.baths_half                   72.15
flags.is_price_reduced                   69.67
price_reduced_amount                     69.67
description.garage                       45.70
description.stories                      23.57
description.sold_price                   18.01
description.lot_sqft                     14.65
description.baths_full                   10.74
description.year_built                   10.68
description.sqft                         10.60
primary_photo.href                        9.62
photos                                    9.62
description.beds                          8.39
location.coun

### Observations: ###
- Almost no entry has a description name, plan_id or spec_id, virtual_tours or 'advertisers' but it's not clear what that would even mean. I want to see what the non-null entries here even look like
- flags.is_foreclosure likely is a yes/no to indicate if this home has been repossessed. Of general housing stock having a very small proportion being foreclosures (0.6% here) seems reasonable so I am confident the nulls here effectively mean 'this home isn't a foreclosure'
- description.baths_3qtr and baths_full probably indicate the number of these structures in the home (same with Garage) so these NaNs can be 0
- What about beds being null (8.3%) or  sqft being null (10.6%)? replace these with the mean?
- list_price should be removed

### Actions for numerical columns: ###
- list_price:
- property_id:                  Drop, serves no analytical purpose, only for a database identifier
- listing_id:                   Drop, serves no analytical purpose, only for a database identifier
- price_reduced_amount:         Replace NaNs with 0, it is showing how much discount is offered
- source.plan_id:               Drop, Probably also a unique identifier with no analytical purpose
- description.year_built:       ???
- description.baths_3qtr:       Replace with 0
- description.sold_price:       ???
- description.baths_half:       Replace with 0
- description.lot_sqft:         ??? Replace with average
- description.sqft:             Replace with average
- description.baths:            Replace with 0
- description.garage:           Replace with 0
- Description.stories:          Replace with 1 - no such thing as a zero story home, assume it means bungalow / one story
- Description.beds:             ???
- location.address.postal_code: Probably fill in nulls with the postal code which matches the area
- location.address.coord.lon    Drop. Coordinates are not relevant in the combined dataset like they may be in specific cities (e.g. North side nicer than South of a given city) and will distort the model.
- location.address.coord.lat    Drop. Coordinates are not relevant in the combined dataset like they may be in specific cities (e.g. North side nicer than South of a given city) and will distort the model.
- location.contu.flips_code     Drop

In [77]:
df[numerical].head(5)

list_price   property_id    listing_id  price_reduced_amount  \
0    554950.0  9.074431e+09  2.957242e+09               45000.0   
1         NaN  9.424984e+09           NaN                   NaN   
2         NaN  9.479069e+09           NaN                   NaN   
3         NaN  9.879332e+09           NaN                   NaN   
4         NaN  9.521640e+09           NaN                   NaN   

   source.plan_id  description.year_built  description.baths_3qtr  \
0             NaN                  1963.0                     NaN   
1             NaN                     NaN                     NaN   
2             NaN                     NaN                     NaN   
3             NaN                     NaN                     NaN   
4             NaN                     NaN                     NaN   

   description.sold_price  description.baths_full  description.baths_half  \
0                     NaN                     2.0                     NaN   
1                     NaN                     NaN                     NaN   
2                     NaN                     NaN                     NaN   
3                     NaN                     NaN                     NaN   
4                     NaN                     NaN                     NaN   

   description.lot_sqft  description.sqft  description.baths  \
0               10454.0            1821.0                2.0   
1                   NaN               NaN                NaN   
2                   NaN               NaN                NaN   
3                   NaN               NaN                NaN   
4                   NaN               NaN                NaN   

   description.garage  description.stories  description.beds  \
0                 1.0                  NaN               3.0   
1                 NaN                  NaN               NaN   
2                 NaN                  NaN               NaN   
3                 NaN                  NaN               NaN   
4                 NaN                  NaN               NaN   

   location.address.postal_code  location.address.coordinate.lon  \
0                       99801.0                       -134.59372   
1                       99801.0                              NaN   
2                       99801.0                              NaN   
3                       99801.0                              NaN   
4                       99801.0                              NaN   

   location.address.coordinate.lat  location.county.fips_code  
0                         58.36395                        NaN  
1                              NaN                        NaN  
2                              NaN                        NaN  
3                              NaN                        NaN  
4                              NaN                        NaN

In [72]:
df[categorical].shape # there are 30 categorical features

(8191, 30)

### Dealing with Tags

Consider the fact that with tags, there are a lot of categorical variables.
- How many columns would we have if we OHE tags, city and state?
- Perhaps we can get rid of tags that have a low frequency.

In [ ]:
# OHE categorical variables/ tags here
# tags will have to be done manually

### Dealing with Cities

- Sales will vary drastically between cities and states.  Is there a way to keep information about which city it is without OHE?
- Could we label encode or ordinal encode?  Yes, but this may have undesirable effects, giving nominal data ordinal values.
- What we can do is use our training data to encode the mean sale price by city as a feature (a.k.a. Target Encoding)
    - We can do this as long as we ONLY use the training data - we're using the available data to give us a 'starting guess' of the price for each city, without needing to encode city explicitly
- If you replace cities or states with numerical values (like the mean price), make sure that the data is split so that we don't leak data into the training selection. This is a great time to train test split. Compute on the training data, and join these values to the test data
- Note that you *may* have cities in the test set that are not in the training set. You don't want these to be NA, so maybe you can fill them with the overall mean

In [ ]:
# perform train test split here
# do something with state and city

## Extra Data - STRETCH

> This doesn't need to be part of your Minimum Viable Product (MVP). We recommend you write a functional, basic pipeline first, then circle back and join new data if you have time

> If you do this, try to write your downstream steps in a way it will still work on a dataframe with different features!

- You're not limited to just using the data provided to you. Think/ do some research about other features that might be useful to predict housing prices. 
- Can you import and join this data? Make sure you do any necessary preprocessing and make sure it is joined correctly.
- Example suggestion: could mortgage interest rates in the year of the listing affect the price? 

In [ ]:
# import, join and preprocess new data here

## EDA/ Visualization

Remember all of the EDA that you've been learning about?  Now is a perfect time for it!
- Look at distributions of numerical variables to see the shape of the data and detect outliers.    
    - Consider transforming very skewed variables
- Scatterplots of a numerical variable and the target go a long way to show correlations.
- A heatmap will help detect highly correlated features, and we don't want these.
    - You may have too many features to do this, in which case you can simply compute the most correlated feature-pairs and list them
- Is there any overlap in any of the features? (redundant information, like number of this or that room...)

In [ ]:
# perform EDA here

## Scaling and Finishing Up

Now is a great time to scale the data and save it once it's preprocessed.
- You can save it in your data folder, but you may want to make a new `processed/` subfolder to keep it organized